File: LC/[1696] Jump Game VI - DP - sliding window max.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        N = len(nums)
        # dp[ii]=max(dp[jj])+nums[ii] where jj in range(max(ii-k, 0), ii)
        # we only care about the max in a sliding window -> sliding window max num pattern
        dq = deque()
        dq.append((nums[0], 0))
        dp=nums[0]
        for ii in range(1, N):
            # evict the oldest
            if dq and ii-dq[0][1] > k: dq.popleft()
           
            dp=dq[0][0]+nums[ii] # calc current dp

            # purge the prior smaller ones
            while dq and dp >= dq[-1][0]: dq.pop()

            dq.append((dp, ii))

        return dp

File: LC/[1696] Jump Game VI - DP - sliding window max.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    static final class Pair {
        int val, idx;
        Pair(int v, int i) { val = v; idx = i; }
    }

    public int maxResult(int[] nums, int k) {
        int n = nums.length;
        Deque<Pair> dq = new ArrayDeque<>();
        dq.addLast(new Pair(nums[0], 0));

        int dp = nums[0];

        for (int i = 1; i < n; i++) {
            // evict outside window
            while (!dq.isEmpty() && i - dq.peekFirst().idx > k) {
                dq.pollFirst();
            }

            dp = dq.peekFirst().val + nums[i];

            // maintain decreasing by dp value
            while (!dq.isEmpty() && dp >= dq.peekLast().val) {
                dq.pollLast();
            }

            dq.addLast(new Pair(dp, i));
        }

        return dp;
    }
}